app/controllers/auth_services_controller.rb

```diff
@@ -61,6 +69,10 @@ def failure
 
   private
 
+  def referer_or_dashboard_path
+    session[:referer_path] || dashboard_path
+  end
+
```

Comment:
For the redirect, this is where we see if a referer path was set above, then we redirect to that page, otherwise we go to the dashboard.